nutils/function.py

```diff
@@ -699,10 +697,29 @@ def lower(self, args: LowerArgs) -> evaluable.Array:
         add_points_shape = tuple(map(evaluable.asarray, self.shape[:self._npointwise]))
         points_shape = args.points_shape + add_points_shape
         coordinates = {space: evaluable.Transpose.to_end(evaluable.appendaxes(coords, add_points_shape), coords.ndim-1) for space, coords in args.coordinates.items()}
-        return _CustomEvaluable(type(self).__name__, self.evalf, self.partial_derivative, evalargs, self.shape[self._npointwise:], self.dtype, self.spaces, types.frozendict(self._arguments), LowerArgs(points_shape, types.frozendict(args.transform_chains), types.frozendict(coordinates)))
-
-    @classmethod
-    def evalf(cls, *args: Any) -> numpy.ndarray:
+        evalf = self.evalf
+        partial_derivative = self.partial_derivative
+        if getattr(self.evalf, '__nutils_hash__', None) is None:
+            warnings.deprecation(f'`{type(self).__name__}.evalf()` does not implement `nutils.types.nutils_hash()`. This will be mandatory in Nutils 10. Try using `nutils.types.hashable_function()`.')
+            evalf = types.hashable_function(f'{type(self).__name__}.evalf')(evalf)
+        if getattr(self.partial_derivative, '__nutils_hash__', None) is None:
+            warnings.deprecation(f'`{type(self).__name__}.partial_derivative()` does not implement `nutils.types.nutils_hash()`. This will be mandatory in Nutils 10. Try using `nutils.types.hashable_function()`.')
+            partial_derivative = types.hashable_function(f'{type(self).__name__}.partial_derivative')(partial_derivative)
+        return _CustomEvaluable(
+            type(self).__name__,
+            evalf,
+            partial_derivative,
+            evalargs,
+            self.shape[self._npointwise:],
+            self.dtype,
+            self.spaces,
+            types.frozendict(self._arguments),
+            LowerArgs(points_shape, types.frozendict(args.transform_chains), types.frozendict(coordinates)),
+        )
+
+    @types.hashable_function('NotImplemented')
+    @staticmethod
```

Comment:
staticmethod is implied by hashable_function

Reply:
fixed